tests/components/airos/conftest.py

```diff
@@ -32,15 +29,28 @@ def mock_setup_entry() -> Generator[AsyncMock]:
 
 
 @pytest.fixture
-def mock_airos_client(ap_fixture: dict[str, Any]):
+def mock_airos_client(
+    request: pytest.FixtureRequest, ap_fixture: AirOSData
+) -> Generator[AsyncMock]:
     """Fixture to mock the AirOS API client."""
-    with patch(
-        "homeassistant.components.airos.AirOS", autospec=True
-    ) as mock_airos_class:
-        mock_client_instance = mock_airos_class.return_value
-        mock_client_instance.login.return_value = True
-        mock_client_instance.status.return_value = ap_fixture
-        yield mock_airos_class
+    mock_airos = AsyncMock()
+    mock_airos.status.return_value = ap_fixture
+
+    if hasattr(request, "param"):
+        mock_airos.login.side_effect = request.param
+    else:
+        mock_airos.login.return_value = True
+
+    with (
+        patch(
+            "homeassistant.components.airos.config_flow.AirOS", return_value=mock_airos
+        ),
+        patch(
+            "homeassistant.components.airos.coordinator.AirOS", return_value=mock_airos
+        ),
+        patch("homeassistant.components.airos.AirOS", return_value=mock_airos),
+    ):
+        yield mock_airos
```

Comment:
```suggestion
    with (
        patch(
            "homeassistant.components.airos.config_flow.AirOS", autospec=True
        ) as mock_airos,
        patch(
            "homeassistant.components.airos.coordinator.AirOS", new=mock_airos
        ),
        patch("homeassistant.components.airos.AirOS", new=mock_airos),
    ):
        client = mock_airos.return_value
        client.status.return_value = ap_fixture
        client.login.return_value = True
        yield client
```
Whenever you need the mock to have a side effect, just add it using `mock_airos_client.login.side_effect = x`